### DISCUSSION TITLE
Self Hosting the webapp?

### DISCUSSION THREAD

**Discussion by Albonycal** [tldraw/tldraw-v1#547] [Q&A]
How to selfhost the webapp?

**Comment by steveruizok** [comment]
You can embed the <Tldraw> component in any React app. You could self-host the webapp (in the `apps/www` folder) though it certain features (analytics, multiplayer) would require environment variables which are private.

**Reply by githubkoma** [reply]
ok what i've found out so far, in my setup:

1) File upload (see below)
2) Regarding Collaboration Feature:
- Example for the website that is being started with ```yarn start:www```
- register, sign in, and obtain your project keys at https://liveblocks.io
  - then, in ```<tldraw dir>/apps/www/.env.local``` add:
  ```
  LIVEBLOCKS_SECRET_KEY=sk_xyz_abc
  NEXT_PUBLIC_LIVEBLOCKS_PUBLIC_API_KEY=pk_xyz_abc
  ```
- then start with ```yarn start:www```
- open http://localhost:3000
- Tap "People" Button -> "Copy to Multiplayer Project"
- -> et voila, you should be in a multiplayer session like http://localhost:3000/r/2acb-4d7c-4d7c-2acb-a6d7a285b728

(1) AWS API Key seems to be needed, dont know if its a "normal" S3 Bucket, i am no AWS guy... 
  ```.env.local``` parameters should to be: 
  ```
  TL_AWS_ACCESS_KEY=xyz
  TL_AWS_SECRET_KEY=abc
  ```

3) More
3.1) Like how to change the "Create a multiplayer project" Button is an <a> https://tldraw.com/r
3.x) anything else??

Please, for 1 + 2 + 3 etc, if there is anything to add, please add more information :-)

P.S.: The basic dev workflow for collab testing was: https://gist.github.com/githubkoma/3d1e4fec6191e215be0323ccb40ce8fc

**Comment by nileshtrivedi** [comment]
It would be great to have a Dockerfile in the repo that builds the app and exposes an HTTP port, making it super-easy to deploy either locally or in containerized environments (ECS/Heroku/CapRover etc)

**Reply by steveruizok** [reply]
I don't have any experience with Docker myself, so it's likely I'm not be seeing the opportunities here. Would you mind expanding on why someone want to host the app separately? Or is it more for local development?

Another thought: the tldraw.com site itself depends on environment variables / api keys that I wouldn't be able to distribute; so a hosted version would likely not work (at least, not the multiplayer parts).

**Reply by nileshtrivedi** [reply]
Companies like to deploy tools on their internal VPN network. They'll use something like AWS Elastic Container Service for running open-source webapps for internal usage. This need is also shared by users who like to self-host things. I personally use CapRover which uses Docker API, LetsEncrypt+Nginx integration for easy mapping to public domains on HTTPS and nice UI for configuring env vars or accessing logs.

The other use-case for Docker is ease of development. Docker has an alpha-status feature called Dev Environments: https://docs.docker.com/desktop/dev-environments/
This allows anybody to clone the repo, launch Dev environment, and launch VS Code integrated. All without having to install the entire stack (NodeJS etc) on their local machine.

**Reply by nileshtrivedi** [reply]
I can take a stab at writing a Dockerfile for `apps/www`.

**Reply by nileshtrivedi** [reply]
Alright, I tried this:

```
FROM node:16

WORKDIR /tldraw​

COPY package.json /tldraw​

RUN npm install

COPY . /tldraw​

EXPOSE 3000

CMD npm run dev
```

But ran into strange Typescript linting errors: 

```
#10 7.819 ./components/MultiplayerEditor.tsx
#10 7.819 1:1  Error: Definition for rule '@typescript-eslint/no-non-null-assertion' was not found
```

**Reply by nileshtrivedi** [reply]
@steveruizok IMHO, API keys and env vars should be documented. It should be up to the user then to procure those for their private deployment. For eg: one of my open-source apps uses Auth0. I don't share my Auth0 keys obviously, but that's not a showstopper for those self-hosting because they can obtain their own Auth0 keys.

**Reply by steveruizok** [reply]
Thanks for the explanation! Sounds like something we should try. 

I'd also be happy to create a minimal deployment that doesn't have things like sponsor-checking, analytics and multiplayer, which are the only parts where tokens are required. The entire "app" is in the @tldraw/tldraw package, so even a minimal React app that embeds the Tldraw component from that package will have all of the app's features.

That error is likely due to the .eslintrc from the root directory. The project always builds from the root, and is using yarn workspaces, so we may need to move over some of those config files if trying to separate out the www folder.

**Reply by swami-at-viasat** [reply]
i tried to yarn the app in docker using this Dockerfile but playwright gave an ssl error when downloading/building. i don't have experience with react/yarn/playwright, so unfortunately didn't get very far with google-searching:
```
FROM ubuntu:focal

RUN apt update

# install nvm
RUN cd ${HOME} && \
    curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.1/install.sh | bash

# install nodejs, packages
RUN . ${HOME}/.bashrc && \
    nvm install --lts && \
    npm install -g yarn

# install tldraw
RUN cd ${HOME} && \
    git clone https://github.com/tldraw/tldraw

RUN cd ${HOME}/tldraw && \
    . ${HOME}/.bashrc && \
    npm config set strict-ssl false -g && \
    yarn config set "strict-ssl" false -g && \
    yarn

RUN cd ${HOME}/tldraw && \
    . ${HOME}/.bashrc && \
    yarn start
```

**Comment by Albonycal** [comment]
+1

**Comment by fabianmu** [comment]
I just managed to self-hosted this on Vercel with the following outline:
 
Create a new project in vercel and change the settings to:

- Framework Preset: `next.js`, obviously however somehow this did not automatically got selected on deployment, so worth checking!  
- BUILD COMMAND: `yarn build:www`  
- OUTPUT DIRECTORY: `apps/www/.next`  

Then run `vercel` in the root of your local code to deploy to vercel and chose the new project or use vercel to connect the project to a GitHub repo.

**Comment by prologic** [comment]
Do you have any Docker images of tldraw I can spin up in my own infra?

**Comment by prologic** [comment]
I _tried_ to quickly put together a `Dockerfile` to build na image from:

```#!Dockerfile
FROM node:current-alpine3.14 as build

RUN npm install -g next

WORKDIR /usr/src/app

COPY package.json yarn.lock ./
RUN yarn
COPY . ./
RUN yarn build

FROM nginx:alpine

COPY --from=build /usr/src/app/build /usr/share/nginx/html

EXPOSE 80

CMD ["nginx", "-g", "daemon off;"]

```

But I've run into some build issues I don't understand. This has less to do with Docker and more to do with how you even build a production version of this app in the first place -- Which there seems to be no documentation for :/

Help 🤗

<img width="1242" alt="Screen Shot 2022-02-20 at 16 21 11" src="https://user-images.githubusercontent.com/1290234/154831298-587f8cbb-a5f8-4cfb-b621-efc8cfd6e44e.png">

**Comment by swami-at-viasat** [comment]
this Dockerfile worked for me, i started from ubuntu:focal and installed nvm and yarn:
```
FROM ubuntu:focal

# update local package databases
RUN apt update

# install some apps
RUN apt install -y curl \
    git

# create a regular account
ENV TLDRAW_USER tldraw
RUN useradd -m ${TLDRAW_USER}

# switch to the new user
USER ${TLDRAW_USER}

# install nvm
RUN cd /home/${TLDRAW_USER} && \
    curl -O https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.1/install.sh && \
    . /home/${TLDRAW_USER}/.profile && \
    NVM_DIR=/home/${TLDRAW_USER}/.nvm PROFILE=/home/${TLDRAW_USER}/.profile /bin/bash ./install.sh

# install nodejs
RUN . /home/${TLDRAW_USER}/.profile && \
    cd /home/${TLDRAW_USER} && \
    nvm install --lts

# install node packages
RUN . /home/${TLDRAW_USER}/.profile && \
    cd /home/${TLDRAW_USER} && \
    npm install --upgrade -g yarn

# clone tldraw
RUN cd /home/${TLDRAW_USER} && \
    git clone https://github.com/tldraw/tldraw

# build tldraw
RUN . /home/${TLDRAW_USER}/.profile && \
    cd /home/${TLDRAW_USER}/tldraw && \
    yarn install --network-timeout 3600000 --non-interactive

COPY yarn_start.sh /home/${TLDRAW_USER}/tldraw/

ENTRYPOINT /home/${TLDRAW_USER}/tldraw/yarn_start.sh
```

yarn_starth.sh is just:
```
#!/usr/bin/env bash

USER=tldraw
. /home/${USER}/.profile
cd /home/${USER}/tldraw
yarn start $@
```

**Reply by prologic** [reply]
I'm not really sure what you did differently to me 🤔 What's the key difference here?

**Reply by prologic** [reply]
I have zero experience with NodeJS/yarn

**Reply by so2liu** [reply]
Will this lose "collaborate" feature?

**Comment by matbgn** [comment]
So do I understand correctly, if I self-host tldraw I want be able to achieve collaborative editing?

**Reply by jreoka** [reply]
Was wondering this as well

**Comment by prologic** [comment]
To answer some recent questions from @so2liu  @matbgn and @DillFrescott the real-time collaboration features of tldraw seem to rely on [Liveblocks](https://liveblocks.io/):

-  [Build amazing real‑time collaborative products - Liveblocks](https://liveblocks.io/)

This in my books makes tldraw unsuitable for self-hosting as one of its more useful features relies on yet another Cloud / SaaS run by yet another company that probably won't respect your data privacy or security.

Someone will have to fork tldraw figure out how to replace the use of LiveBlocks with something else. HedgeDoc have a similar set of features and don't need to rely on an external (random) Cloud/SaaS service.

**Reply by matbgn** [reply]
This situation is known and a reflection is in progress by @steveruizok to address this aspect. 

The basic idea was to use yjs like @nimeshnayaju did on his fork (POC) https://github.com/nimeshnayaju/yjs-tldraw

Feel free to address the remaining issues encounters by the POC and make a Pull Request here.

**Reply by prologic** [reply]
Thank you! 🙏

**Reply by ocdtrekkie** [reply]
Also watching this topic. (Hi @prologic, funny finding you here...)

**Comment by inquam** [comment]
I see a lot of interesting conversations here. Do we have any update on a working Docker file for self-hosting with working collaboration and storage?

**Reply by prologic** [reply]
Nope, I've tried several times and haven't succeeded.
